vncviewer/touch.cxx

```diff
@@ -0,0 +1,324 @@
+extern void dbgEvent(const XEvent& event);
+#endif
+
+static int handleTouchEvent(void *event, void *data)
+{
+#if defined(WIN32)
+  MSG *msg = (MSG*)event;
+
+  // Don't proceed if this isn't a touch device
+  if (!GetSystemMetrics(SM_DIGITIZER)) {
+    return 0;
+  }
+
+  // If window is registered as touch we can not receive gestures
+  if (IsTouchWindow(msg->hwnd, NULL)) {
+    vlog.error(_("Window is registered for touch instead of gestures"));
+    return 0;
+  }
+
+  // Trigger on the first WM_PAINT event. We can't use WM_CREATE events
+  // since FLTK's system handlers trigger before WndProc.
+  // WM_CREATE events are sent directly to WndProc.
+  if (msg->message == WM_PAINT && handlers.count(msg->hwnd) == 0) {
+    handlers[msg->hwnd] = new win32TouchHandler(msg->hwnd);
+    // We need a special hook-in to be able to handle gestures.
+    // Again, we can't rely on FLTK's system handler won't work here.
+    // WM_GESTURE events are generated by WndProc for unhandled WM_POINTER
+    // events. FLTK's system handlers trigger before WndProc.
+    if (!SetWindowSubclass(msg->hwnd, &win32_gestureProc, 1, 0)) {
+      vlog.error(_("Failed to insert gesture handler on window"));
+      return 1;
+    }
+  }
+#elif !defined(WIN32) && !defined(__APPLE__)
+  XEvent *xevent = (XEvent*)event;
+
+
+  if ((xevent->type == ButtonPress) ||
+      (xevent->type == ButtonRelease)) {
+    vlog.debug("Got incoming event:");
+    dbgEvent(*xevent);
+  }
+
+  if (xevent->type == MapNotify) {
+    handlers[xevent->xmap.window] = new XInputTouchHandler(xevent->xmap.window);
+
+    // Fall through as we don't want to interfere with whatever someone
+    // else might want to do with this event
+
+  } else if (xevent->type == UnmapNotify) {
+    delete handlers[xevent->xunmap.window];
+    handlers.erase(xevent->xunmap.window);
+  } else if (xevent->type == DestroyNotify) {
+    delete handlers[xevent->xdestroywindow.window];
+    handlers.erase(xevent->xdestroywindow.window);
+  } else if (xevent->type == GenericEvent) {
+    if (xevent->xgeneric.extension == xi_major) {
+      XIDeviceEvent *devev;
+
+      if (!XGetEventData(fl_display, &xevent->xcookie)) {
+        vlog.error(_("Failed to get event data for X Input event"));
+        return 1;
+      }
+
+      devev = (XIDeviceEvent*)xevent->xcookie.data;
+
+      if (handlers.count(devev->event) == 0) {
+        vlog.error(_("X Input event for unknown window"));
+        XFreeEventData(fl_display, &xevent->xcookie);
+        return 1;
+      }
+
+      dynamic_cast<XInputTouchHandler*>(handlers[devev->event])->processEvent(devev);
+
+      XFreeEventData(fl_display, &xevent->xcookie);
+
+      return 1;
+    }
+  }
+#endif
+
+  return 0;
+}
+
+void enable_touch()
+{
+#if defined(WIN32)
+  // Only write to debug since this will trigger on all non-touch devices
+  if (!GetSystemMetrics(SM_DIGITIZER))
+    vlog.debug("The 'Tablet PC Input' service isn't running. Touch gestures will not be supported.");
```

Comment:
What if a touch device is plugged in later?

Reply:
I rewrote most of this code in 7fcb1048d6292357c133b93b93d6c1fa03a97ad6